File: day21/Part1.scala
import scala.io.Source
import scala.collection.mutable.HashMap
import scala.collection.mutable.HashSet
import scala.collection.mutable.ListBuffer

object Part1 {
  def main(args: Array[String]): Unit = {
    val filename = "input"
    val file = Source.fromFile(filename)
    val lines = file.getLines().toList
    val sources = HashMap[String, ListBuffer[String]]()
    val safeIngs = HashSet[String]()
    for(line <- lines) {
      val split = line.split(" \\(contains ")
      val ingredients = split(0).split(" ")
      safeIngs ++= ingredients
      val allergens = split(1).replace(")", "").split(", ")
      for(allergen <- allergens) {
        if(sources.contains(allergen)) sources(allergen) = sources(allergen).filter(ingredients.contains(_))
        else sources += (allergen -> ingredients.to(ListBuffer))
      }
    }
    for((key, value) <- sources) safeIngs --= value
    var count = 0
    val words = lines.map(_.split(' ')).flatten
    for(ing <- safeIngs) count += words.count(_ == ing)
    println(count)
  }
}

File: year2021/day05/part1.rb
lines = File.read('input').each_line.map do |line|
  line.split(' -> ').map { |el| el.split(',').map(&:to_i) } 
end
lines.select! { |line| line[0][0] == line[1][0] || line[0][1] == line[1][1] }
ocean = Array.new(1000) { Array.new(1000, 0) }
lines.each do |line|
  x1 = line[0][0]
  y1 = line[0][1]
  x2 = line[1][0]
  y2 = line[1][1]
  ([x1, x2].min..[x1,x2].max).each do |x|
    ([y1, y2].min..[y1, y2].max).each do |y|
      ocean[y][x] = ocean[y][x] + 1
    end
  end
end
puts ocean.flatten.count { |x| x > 1 }
